src/core/namespace.c

```diff
@@ -2245,32 +2255,73 @@ int setup_namespace(const NamespaceParameters *p, char **error_path) {
         if (r < 0)
                 return r;
 
-        if (p->tmp_dir) {
-                bool ro = streq(p->tmp_dir, RUN_SYSTEMD_EMPTY);
+        if (p->private_tmp_as_tmpfs) {
+                _cleanup_free_ char *tmpfs_dir = NULL, *tmp_dir = NULL, *var_tmp_dir = NULL;
+                MountEntry *tmpfs_entry, *tmp_entry, *var_tmp_entry;
 
-                MountEntry *me = mount_list_extend(&ml);
-                if (!me)
+                tmpfs_dir = strdup(p->private_tmp_dir);
+                tmp_dir = path_join(p->private_tmp_dir, "tmp");
+                var_tmp_dir = path_join(p->private_tmp_dir, "var-tmp");
+                if (!tmpfs_dir || !tmp_dir || !var_tmp_dir)
                         return log_oom_debug();
 
-                *me = (MountEntry) {
+                tmpfs_entry = mount_list_extend(&ml);
+                tmp_entry = mount_list_extend(&ml);
+                var_tmp_entry = mount_list_extend(&ml);
+                if (!tmpfs_entry || !tmp_entry || !var_tmp_entry)
+                        return log_oom_debug();
+
+                *tmpfs_entry = (MountEntry) {
+                        .path_malloc = TAKE_PTR(tmpfs_dir),
+                        .mode = MOUNT_TMPFS,
+                        .options_const = "mode=01700" NESTED_TMPFS_LIMITS,
```

Comment:
hmm, i'd make this 0700 btw. eve if this is ultimately not visible i think a dir with sgid is weird if not used like that.